Add MiniCPM3

bzantium opened the pull request:
# What does this PR do?

Adds native support for the **MiniCPM3** architecture from OpenBMB
([`openbmb/MiniCPM3-4B`](https://huggingface.co/openbmb/MiniCPM3-4B)).

MiniCPM3 combines:

- **Multi-head Latent Attention (MLA)** from DeepSeek-V2, with the original cos/sin rotary embedding convention preserved on the rotary part of the query/key heads.
- A standard SwiGLU MLP (no MoE).
- Three scalar scaling factors:
  - `scale_emb` — multiplier on input embeddings.
  - `scale_depth / sqrt(num_hidden_layers)` — multiplier on residual connections.
  - `hidden_size / dim_model_base` — divisor on hidden states before the LM head.

When `scale_depth` / `dim_model_base` are not provided, they default to no-op values (`sqrt(num_hidden_layers)` and `hidden_size` respectively).

Implemented as a single `modular_minicpm3.py` deriving from `LlamaConfig` / `LlamaModel`, with a custom `MiniCPM3Attention` for MLA. The model reuses the upstream `LlamaTokenizer` (no new tokenizer files); `minicpm3` is added to `MODELS_WITH_INCORRECT_HUB_TOKENIZER_CLASS` so the auto loader falls back to `TokenizersBackend`. `AutoConfig` / `AutoModel` / `AutoModelForCausalLM` / `AutoModelForSequenceClassification` mappings, docs, and tests are wired up.

Fixes #41115.

A second, more recent PR (#45613) covers the same model. This PR (#41116) is the original; the implementations differ on the RoPE convention (this one preserves the original cos/sin RoPE; #45613 uses DeepSeek-V2's complex RoPE) and on a few config defaults.

AI assistance was used during the rebase/refactor; the changes have been reviewed and tested locally end-to-end (see follow-up comment for the example run on `openbmb/MiniCPM3-4B`).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. — #41115
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @Cyrilvallez

bzantium commented:
@ArthurZucker please check this out! I've tested with MiniCPM3-4B and it worked well.

bzantium committed:
Add MiniCPM3

Adds support for the OpenBMB MiniCPM3 architecture (e.g.
`openbmb/MiniCPM3-4B`). MiniCPM3 combines Multi-head Latent
Attention (MLA) from DeepSeek-V2 with a standard SwiGLU MLP and
three scalar scaling factors that govern signal flow:

- `scale_emb` scales input embeddings.
- `scale_depth / sqrt(num_hidden_layers)` scales residual
  connections.
- `hidden_size / dim_model_base` scales hidden states before the
  language-model head.

The implementation follows the modular pattern, deriving the config
and model from `LlamaConfig` / `LlamaModel`. The MLA attention
keeps the cos/sin rotary convention used by the original
implementation, registered classes are wired into the auto
mappings, and `tokenizer_auto` is updated so the upstream tokenizer
falls back to the tokenizers backend.

Verified with `openbmb/MiniCPM3-4B` end-to-end: weights load
without missing/mismatched keys and greedy generation produces a
coherent reply via `attn_implementation="eager"`.

Co-authored-by: Aladdin Aliyev <213189260+aliyevaladddin@users.noreply.github.com>

bzantium committed:
Default MiniCPM3 scalings to no-ops when unspecified

`scale_depth=1.0` and `dim_model_base=1` make the residual and
logit scalings collapse, which prevented the small tester model
from training and broke the `test_training_overfit` CI job.

Mirror the original MiniCPM3 defaults: when not provided,
`scale_depth` falls back to `sqrt(num_hidden_layers)` and
`dim_model_base` falls back to `hidden_size`, so both factors
become exact no-ops. Drop the explicit overrides in
`MiniCPM3ModelTester` so the small test config inherits the new
defaults.

The real `openbmb/MiniCPM3-4B` config is unaffected: it sets
all three scalars explicitly.

Co-authored-by: Aladdin Aliyev <213189260+aliyevaladddin@users.noreply.github.com>

bzantium commented:
End-to-end check on `openbmb/MiniCPM3-4B`:

```python
import torch
from transformers import AutoTokenizer, MiniCPM3ForCausalLM

tokenizer = AutoTokenizer.from_pretrained("openbmb/MiniCPM3-4B")
model = MiniCPM3ForCausalLM.from_pretrained(
    "openbmb/MiniCPM3-4B",
    dtype=torch.float16,
    device_map="auto",
    attn_implementation="eager",
).eval()

messages = [{"role": "user", "content": "Tell me a fun fact about the Eiffel Tower."}]
text = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
inputs = tokenizer(text, return_tensors="pt").to(model.device)
out = model.generate(**inputs, max_new_tokens=80, do_sample=False)
print(tokenizer.decode(out[0], skip_special_tokens=True))
```

Output:

```
user
Tell me a fun fact about the Eiffel Tower.
 assistant
The Eiffel Tower was originally intended to be a temporary structure, designed for the 1889 World's Fair to celebrate the 100th anniversary of the French Revolution. It was meant to be a symbol of the Industrial Revolution and a demonstration of France's engineering prowess. However, it was so popular that it was allowed to stand after the Fair. It
```

bzantium commented:
@aliyevaladddin Sorry I missed your email earlier. I just rebased this PR, would be happy to add you as coauthor since your message got me to come back to it.

bzantium commented:
@ArthurZucker @Cyrilvallez when you have a moment, would appreciate a review on this. Thanks!

aliyevaladddin commented:
Thanks for the rebase, @bzantium! I'm honored to be added as a co-author. I'm really excited to see MiniCPM3 support getting closer to being merged. Let me know if there's anything else I can help with regarding testing or further updates!

vasqu commented:
Will try to take a look today :hugs:

vasqu commented on docs/source/en/model_doc/minicpm3.md:
```diff
@@ -0,0 +1,67 @@
+<!--Copyright 2025 The HuggingFace Team. All rights reserved.
```
```suggestion
<!--Copyright 2026 The HuggingFace Team. All rights reserved.
```

bzantium replied:
Done, bumped to 2026.

vasqu commented on docs/source/en/model_doc/minicpm3.md:
```diff
@@ -0,0 +1,67 @@
+rendered properly in your Markdown viewer.
+
+-->
+*This model was released on 2024-09-05 and added to Hugging Face Transformers on 2026-04-29.*
```
Pretty sure the wording was slightly updated so please sync with main and recheck

bzantium replied:
Rechecked against main: the doc-builder note and the 'released on / added on' line already match the current wording, so I just bumped the copyright year.

vasqu commented on docs/source/en/model_doc/minicpm3.md:
````diff
@@ -0,0 +1,67 @@
+## Usage tips
+
+```python
+from transformers import AutoTokenizer, MiniCPM3ForCausalLM
````
please use automodelforcausallm - its more user friendly and oftentimes reveals if we missed some mapping

bzantium replied:
Done, switched the usage example to AutoModelForCausalLM.

vasqu commented on src/transformers/models/minicpm3/__init__.py:
```diff
@@ -0,0 +1,29 @@
+# Copyright 2025 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2026 The HuggingFace Team. All rights reserved.
```

bzantium replied:
Done.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+# Copyright 2025 The OpenBMB Team and the HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2026 The OpenBMB Team and the HuggingFace Inc. team. All rights reserved.
```

bzantium replied:
Done.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+    scale_depth (`int` or `float`, *optional*):
+        Multiplier for residual connections; the effective scaling is `scale_depth / sqrt(num_hidden_layers)`.
+        If `None`, defaults to `sqrt(num_hidden_layers)` (no-op scaling).
+    dim_model_base (`int`, *optional*):
```
Is this actually used in any of the models then if we default to hidden size == no effect? If its not really used, we could remove it imo

bzantium replied:
These are not identity operations for the released checkpoint: openbmb/MiniCPM3-4B sets scale_emb=12, scale_depth=1.4, dim_model_base=256, so the input to the LM head is divided by hidden_size / dim_model_base = 2560 / 256 = 10. The defaults only collapse to identity when left unset (for example a randomly initialised tiny test model), so I kept dim_model_base and exposed the logit scaling as a config property (see below).

vasqu replied:
But the default values should match to those of `openbmb/MiniCPM3-4B` (the ckpt in auto docs). Im fine with defaulting to a no op if it is explicitly set to None

bzantium replied:
Good point — updated the class defaults to match `openbmb/MiniCPM3-4B` (`scale_emb=12`, `scale_depth=1.4`, `dim_model_base=256`). The no-op behaviour is now only triggered when a field is explicitly set to `None` (e.g. a tiny randomly-initialised test model), as you suggested.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+    v_head_dim: int | None = None
+    scale_emb: int | float = 1.0
+    scale_depth: int | float | None = None
+    dim_model_base: int | None = None
```
```suggestion
    vocab_size: int = 73448
    hidden_size: int = 2560
    intermediate_size: int = 6400
    num_hidden_layers: int = 62
    num_attention_heads: int = 40
    num_key_value_heads: int | None = 40
    max_position_embeddings: int = 32768
    initializer_range: float = 0.1
    rms_norm_eps: float = 1e-5
    tie_word_embeddings: bool = True
    kv_lora_rank: int = 256
    q_lora_rank: int | None = 768
    qk_nope_head_dim: int = 64
    qk_rope_head_dim: int = 32
    v_head_dim: int | None = None
    scale_emb: int | float = 1.0
    scale_depth: int | float | None = None
    dim_model_base: int | None = None
```
we can remove a few things here since they are already defined in llama with the same values

bzantium replied:
Done, removed the fields that match LlamaConfig defaults and kept only the ones that differ plus the MLA and scaling extras (also dropped the now unused RopeParameters import).

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+
+    def __post_init__(self, **kwargs):
+        # In MLA the per-head dim used by RoPE is the rotary part, not `hidden_size / num_attention_heads`.
+        self.head_dim = self.qk_rope_head_dim
```
Ig we need this because of the remote configs?

bzantium replied:
Same as DeepseekV2Config: in MLA the per head dim used by RoPE is the rotary part (qk_rope_head_dim), so head_dim has to be set here for the rope and cache shapes. Kept it, and it is now consistent with the inherited DeepseekV2Attention.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+            self.dim_model_base = self.hidden_size
+        super().__post_init__(**kwargs)
+
+    def validate_architecture(self):
```
Seems like we dont overwrite in e.g. dsv3. Any reason we need it here?

bzantium replied:
Removed the override so the standard Llama divisibility check applies, matching dsv2 and dsv3. The released config (2560 % 40 == 0) and the test configs pass, so it was not needed.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+    pass
+
+
+class MiniCPM3Attention(nn.Module):
```
We can still inherit from some model like dsv3 or dsv2 no? There should be some big similarities

You can delete attributes with `del self.attr` if you dont want to inherit or override/add by just declaring like usual `self.attr = ..`

vasqu replied:
Also noticing doesnt this match dsv2 exactly? Why not inherit from there?

bzantium replied:
Good point. The __init__ is indeed identical to DeepseekV2Attention and forward only differs in the RoPE call: MiniCPM3 keeps the original cos/sin convention (apply_rotary_pos_emb) whereas DeepSeek-V2 uses the complex interleaved apply_rotary_emb. I now inherit from DeepseekV2Attention and override only forward, which drops the duplicated __init__. Verified the attention output is bit identical before and after the change. The rotary embedding stays Llama based (cos/sin), and the decoder layer cannot be inherited from dsv2 since MiniCPM3 adds the scale_depth residual scaling.

bzantium replied:
Right, the module construction matches dsv2 exactly. Done as discussed in the other thread: MiniCPM3Attention now inherits DeepseekV2Attention and overrides only forward for the cos/sin RoPE.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+        else:
+            self.q_a_proj = nn.Linear(self.hidden_size, config.q_lora_rank, bias=config.attention_bias)
+            self.q_a_layernorm = MiniCPM3RMSNorm(config.q_lora_rank)
+            self.q_b_proj = nn.Linear(config.q_lora_rank, self.num_heads * self.qk_head_dim, bias=False)
```
Is this actually something that happens. Would like to avoid if possible but no worries if not

bzantium replied:
The released checkpoint uses q_lora_rank=768, so the None branch is dead for the real weights. It now comes for free from the inherited DeepseekV2Attention, so it is no longer extra code here.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states * (self.scale_depth / math.sqrt(self.num_hidden_layers))
```
Why not directly calculate the scale as attributes? We dont really need to do this each time and let's add a small comment higlighting the diff

bzantium replied:
Done, precomputed as self.residual_scale in the decoder layer __init__ and reused for both residual branches, with a comment noting the diff from Llama.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+            raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
+
+        if inputs_embeds is None:
+            inputs_embeds: torch.Tensor = self.embed_tokens(input_ids) * self.config.scale_emb
```
same here to highlight

bzantium replied:
Added comments highlighting the embedding, residual and logit scalings versus Llama, and both residual branches now use the precomputed self.residual_scale.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,444 @@
+
+        hidden_states = outputs.last_hidden_state
+        # MiniCPM3 logit scaling: divide hidden states before the LM head.
+        hidden_states = hidden_states / (self.config.hidden_size / self.config.dim_model_base)
```
should be a direct config attribute imo

bzantium replied:
Done, exposed it as a MiniCPM3Config.logits_scaling property (hidden_size / dim_model_base); the forward now divides by self.config.logits_scaling.

vasqu commented on tests/models/minicpm3/test_modeling_minicpm3.py:
```diff
@@ -0,0 +1,109 @@
+# Copyright 2025 The OpenBMB Team and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2026 The OpenBMB Team and The HuggingFace Inc. team. All rights reserved.
```

bzantium replied:
Done.

vasqu commented on tests/models/minicpm3/test_modeling_minicpm3.py:
```diff
@@ -0,0 +1,109 @@
+
+@require_torch
+class MiniCPM3ModelTest(CausalLMModelTest, unittest.TestCase):
+    test_all_params_have_gradient = False
```
```suggestion
```
shouldnt be needed, it mostly happens on moe models

bzantium replied:
Done, removed test_all_params_have_gradient = False. MiniCPM3 is dense, so the MoE oriented flag is not needed and all params get gradients.

vasqu commented on tests/models/minicpm3/test_modeling_minicpm3.py:
```diff
@@ -0,0 +1,109 @@
+
+
+@require_torch
+class MiniCPM3IntegrationTest(unittest.TestCase):
```
please check minimax m2 for proper integration tests: we want actual values (I can adjust for our CI)

bzantium replied:
Reworked the integration tests to the value based Expectations pattern (a logits slice plus exact generated text), following MiniMax-M2. The expected values still need a reference run on the target CI hardware, so I left them as Expectations placeholders that skip until filled. Would appreciate your help adjusting those for CI as you offered.

vasqu replied:
Yep no worries, will adjust for our CI

vasqu reviewed (commented):
Some initial comments from my side 🫡

bzantium committed:
Address review: inherit dsv2 attention, slim config, precompute scalings

- MiniCPM3Attention now inherits DeepseekV2Attention and overrides only
  forward (cos/sin RoPE instead of DeepSeek-V2's complex rotary), removing
  the duplicated MLA __init__. Forward output is numerically identical.
- Drop config fields that match LlamaConfig defaults; keep only the ones
  that differ plus the MLA/scaling extras.
- Remove the validate_architecture override so the standard Llama
  divisibility check applies, matching dsv2/dsv3.
- Precompute the residual depth scaling per layer; expose the logit
  scaling as a config property; add comments highlighting the diffs from
  Llama (embedding, residual, logit scalings).
- Bump copyright years to 2026; use AutoModelForCausalLM in the docs
  example; drop the MoE-oriented test_all_params_have_gradient flag.
- Rework integration tests to the value-based Expectations pattern
  (expected logits/text to be filled from a CI reference run).

bzantium committed:
Merge remote-tracking branch 'origin/main' into feature/#41115

# Conflicts:
#	docs/source/en/_toctree.yml
#	src/transformers/models/auto/modeling_auto.py

bzantium committed:
docs: sync MiniCPM3 model card date line with current wording

Update the date line to the current "published in HF papers on ... and
contributed to ..." format expected by utils/add_dates.py, fixing the
check_repository_consistency failure.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,394 @@
+    }
+
+    model_type = "minicpm3"
+    keys_to_ignore_at_inference = ["past_key_values"]
```
```suggestion
```
should be the same in llama no?

bzantium replied:
Right, `keys_to_ignore_at_inference = ["past_key_values"]` is already inherited from `LlamaConfig`. Removed.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,394 @@
+
+        cos, sin = position_embeddings
+        q_rot, k_rot = apply_rotary_pos_emb(q_rot, k_rot, cos, sin)
+        k_rot = k_rot.expand(*k_pass.shape[:-1], -1)
```
So the forward is essentially dsv3 with the non interleave, correct? Could we add a small comment here to highlight?

bzantium replied:
Correct. Added a comment at the rotary call noting the forward is the same MLA path as DeepSeek-V2/V3, except MiniCPM3 keeps the standard (non-interleaved) cos/sin rotary instead of DeepSeek's complex/interleaved variant.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -0,0 +1,394 @@
+class MiniCPM3DecoderLayer(LlamaDecoderLayer):
+    def __init__(self, config: MiniCPM3Config, layer_idx: int):
+        super().__init__(config, layer_idx)
+        self.self_attn = MiniCPM3Attention(config=config, layer_idx=layer_idx)
```
```suggestion
```
shouldnt be needed, no? Modular should also inherit this properly I think

bzantium replied:
You're right, modular renames the inherited `self_attn` to `MiniCPM3Attention` automatically, so the explicit assignment was redundant. Removed; the generated `modeling_minicpm3.py` is unchanged.

vasqu commented on tests/models/minicpm3/test_modeling_minicpm3.py:
```diff
@@ -0,0 +1,132 @@
+        # adjust per-hardware entries as needed (see `Expectations`).
+        expected_slices = Expectations(
+            {
+                ("cuda", 8): [],  # TODO: fill from a reference run on the target CI hardware
```
Can you at least add some values based on your hw for now?

bzantium replied:
Ran `openbmb/MiniCPM3-4B` on an A100 (bf16) and filled in the `Expectations`:
- logits slice: `[0.765625, 3.640625, -0.189453125, -0.8359375, -0.8359375]`
- generation: `"My favourite condiment is \n[A]. ketchup \n[B]. mustard \n[C"`

Both `test_minicpm3_4b_logits` and `test_minicpm3_4b_generation` pass under `("cuda", 8)`. Feel free to adjust/add per-hardware entries for your CI.

vasqu reviewed (approved):
Just some small comments but overall lgtm. Let me update the CI tomorrow then after you fixed the last comments 🫡

vasqu commented:
We also need to sync with main and fix CI please

bzantium committed:
Merge branch 'main' into feature/#41115

bzantium committed:
Address review: align config defaults, scaled embedding, fill integration values

- Default config to the openbmb/MiniCPM3-4B checkpoint values (scale_emb=12,
  scale_depth=1.4, dim_model_base=256); no-op scaling only when set to None.
- Drop redundant keys_to_ignore_at_inference (inherited from LlamaConfig).
- Move embedding scaling into MiniCPM3ScaledWordEmbedding so input_ids and
  inputs_embeds paths stay consistent (fixes inputs_embeds-vs-input_ids tests).
- Remove redundant self_attn assignment in the decoder layer (modular renames it).
- Highlight the MLA RoPE diff vs DeepSeek-V2/V3 with a comment.
- Fill integration test Expectations with verified A100 (bf16) reference values.

bzantium commented:
@vasqu 

The `security-gate / security-check` failure comes from the "Enforce allowlist" step. The only offending file is `docs/source/en/_toctree.yml`, which the gate rejects because `.yml` is not allowed outside `tests/` (only `.py`, `.txt`, `.md` are permitted there). That `_toctree.yml` entry is required to register MiniCPM3 in the docs navigation, so it cannot be dropped.

This gate only runs for fork PRs, and the failure comment notes that a maintainer can manually approve CI when a finding is a false positive. Could you approve CI when you get a chance? All 11 other changed files pass the allowlist. Thanks!

bzantium committed:
Skip test_sdpa_can_dispatch_on_flash for MiniCPM3 (MLA head dims)

MiniCPM3 uses MLA (inherited from DeepSeek-V2), so the query/key head dim
(qk_nope + qk_rope) differs from the value head dim. PyTorch's flash kernel
requires q, k, v to share the same last dim, so SDPA cannot dispatch on flash.
Skip the test as DeepSeek-V3 already does for the same reason.

vasqu commented on src/transformers/models/minicpm3/modular_minicpm3.py:
```diff
@@ -136,6 +138,21 @@ def logits_scaling(self) -> float:
         return self.hidden_size / self.dim_model_base
 
 
+class MiniCPM3ScaledWordEmbedding(nn.Embedding):
```
Can we inherit from gemma3 for example?

vasqu committed:
fixes for ci + simplify modular a bit

vasqu reviewed (approved):
LGTM, the logits differ quite a bit from 8 to 8.6 so just wanna make sure if you can check that I didnt break anything

vasqu commented:
gentle ping @bzantium

bzantium committed:
Fix repo consistency: allowlist dim_model_base and refresh minicpm3 dates

bzantium committed:
Merge branch 'main' into feature/#41115

bzantium commented:
@vasqu Thanks for the careful check! I verified on an A100 (compute capability 8.0, same hardware the `("cuda", 8)` reference came from) using the current branch, which includes your modular cleanup. Nothing is broken.

**bf16 reproduction (A100, cc 8.0):**
```
slice : [0.765625, 3.640625, -0.189453, -0.835938, -0.835938]
ref   : [0.765625, 3.640625, -0.189453125, -0.8359375, -0.8359375]
diff  : [0, 0, 0, 0, 0]   # exact match
```
So your refactor (inheriting `Gemma3TextScaledWordEmbedding` and dropping the duplicated `MiniCPM3Model.forward`) is numerically a no op. That matches the code too: the new embedding base is identical to the old one, and the removed `forward` was a verbatim copy of the inherited `LlamaModel.forward`.

**On the 8 vs 8.6 gap:** both buckets run bf16 (the config `torch_dtype` is `bfloat16`, and the test loads with `dtype="auto"`), so the difference is purely cross GPU bf16 variation, not a dtype or code change. These particular logits sit near zero and the model is 62 layers deep, so bf16 rounding accumulates and the low magnitude entries wobble. To put a number on it, I also ran fp32 as a ground truth on the same A100:
```
fp32  : [0.535951, 3.610042, -0.174161, -1.174419, -1.174420]
```
The bf16 vs fp32 gap reaches about 0.34, which is larger than the 8.0 vs 8.6 gap of about 0.08. Most importantly, the argmax token is 11213 in bf16 and fp32 alike, so the actual prediction is stable across precision and only the near zero logit tail moves.

Net: your `(8, 6)` bucket is a legitimate per hardware entry and the existing `(8)` reference still reproduces exactly. No regression. Good to go from my side.

vasqu commented:
run-slow: minicpm3

vasqu commented:
Just sanity checking but should be good to go then :saluting_face:

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_41116). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

vasqu commented:
Nice thanks for sticking through and gz on the merge :hugs:

aliyevaladddin commented:
Finaly!
